### ISSUE TITLE
Cookie-Based auth return 401 while the Cookies is valid

### ISSUE THREAD

**Issue by atom992** [couchbase/couchbase-lite-android#817]
I am using couchbase-lite-android:1.2.0 with Couchbase Sync Gateway 1.2.0 on Android 5.0.1.
When I used  couchbase-lite REST API to setup pull and push replication,it works.
but after few hours(the Cookies is valid),the replication doesn't work.
Cookies Check:
```
curl -v -H "Content-Type: application/json" --cookie "SyncGatewaySession=60445a4d3b4089a1ec726976d815c9917bd59cd2" http://10.0.1.1:4984/sgdb/doc1
```
Return the doc1 info.

adb logcat logs shows:
```
D/Sync    ( 8568): com.couchbase.lite.replicator.PullerInternal$12@27f05333 still running, restarting change tracker
D/Sync    ( 8568): com.couchbase.lite.replicator.PullerInternal@395233f0: starting ChangeTracker with since=10656 mode=OneShot
D/Sync    ( 8568): com.couchbase.lite.replicator.PullerInternal@395233f0: started ChangeTracker com.couchbase.lite.replicator.ChangeTracker@71f26f0
D/ChangeTracker( 8568): com.couchbase.lite.replicator.ChangeTracker@71f26f0: Changed tracker asked to start
D/ChangeTracker( 8568): Thread id => 1100
V/ChangeTracker( 8568): com.couchbase.lite.replicator.ChangeTracker@71f26f0: Making request to http://10.0.1.1:4984/sgdb/_changes
E/ChangeTracker( 8568): com.couchbase.lite.replicator.ChangeTracker@71f26f0: Change tracker got error 401
V/ChangeTracker( 8568): com.couchbase.lite.replicator.ChangeTracker@71f26f0: Change tracker run loop exiting
D/ChangeTracker( 8568): com.couchbase.lite.replicator.ChangeTracker@71f26f0: Change tracker in stopped()
D/ChangeTracker( 8568): com.couchbase.lite.replicator.ChangeTracker@71f26f0: Change tracker calling changeTrackerStopped, client: com.couchbase.lite.replicator.PullerInternal@395233f0

```

Couchbase Sync Gateway error logs shows:
```
2016-03-23T10:38:19.857+08:00 HTTP:  #28515: POST /sgdb/_changes
2016-03-23T10:38:19.857+08:00 HTTP: #28515:     --> 401 Login required  (0.7 ms)
2016-03-23T10:38:30.055+08:00 HTTP:  #28516: POST /sgdb/_changes
2016-03-23T10:38:30.055+08:00 HTTP: #28516:     --> 401 Login required  (0.5 ms)

```

HTTP Request/Response shows:
Request:
```
POST /sgdb/_changes HTTP/1.1
Content-Type: application/json
User-Agent: CouchbaseLite/1.2 (1.2.0/3fa390b596672fa39243cb9f3457c276eb53707b)
Accept-Encoding: gzip
Cookie: SyncGatewaySession=60445a4d3b4089a1ec726976d815c9917bd59cd2
Content-Length: 101
Host: 10.0.1.1:4984
Connection: Keep-Alive
Cookie: SyncGatewaySession=60445a4d3b4089a1ec726976d815c9917bd59cd2
Cookie2: $Version=1

{"style":"all_docs","heartbeat":300000,"feed":"normal","active_only":null,"limit":null,"since":10656}
```
Response:
```
HTTP/1.1 401 Unauthorized
Date: Wed, 23 Mar 2016 02:23:11 GMT
Content-Type: application/json
Content-Length: 50
Connection: keep-alive
server: Couchbase Sync Gateway/1.2.0
www-authenticate: Basic realm="Couchbase Sync Gateway"

{"error":"Unauthorized","reason":"Login required"}
```

**Comment by atom992** [comment]
I think I should submit this issue to Couchbase Sync Gateway.

**Comment by atom992** [comment]
I used curl to send the same request to sync gateway, and it return values:
```
curl -v -X POST -H "Content-Type: application/json" --cookie "SyncGatewaySession=60445a4d3b4089a1ec726976d815c9917bd59cd2" http://10.0.1.1:4984/sgdb/_changes -d '{"style":"all_docs","heartbeat":300000,"feed":"normal","active_only":null,"limit":null,"since":10656}'
```
It return the following info:
```
< HTTP/1.1 200 OK
< Date: Wed, 23 Mar 2016 03:26:10 GMT
< Content-Type: application/json
< Content-Length: 37
< Connection: keep-alive
< cache-control: private, max-age=0, no-cache, no-store
< server: Couchbase Sync Gateway/1.2.0
<
{"results":[
],
"last_seq":"10656"}
* Connection #0 to host 10.0.1.1 left intact
```

and Sync Gateway error log shows:
```
10.0.1.101 - - [23/Mar/2016:11:26:38 +0800] "POST /sgdb/_changes HTTP/1.1" 200 20780 "-" "curl/7.43.0"
```

So I am not sure this is cause by Couchabse-lite or Sync Gateway. I will reopen the issue.

**Comment by atom992** [comment]
any update?

**Comment by hideki** [comment]
sorry, this ticket is still in queue.

**Comment by atom992** [comment]
any update?

**Comment by npomfret** [comment]
I get the same issue in the client.  I have a react-native app that calls the CBL rest api and from time to time it returns a 401 along with a www-authenticate header.  

Retrying the request gets around the problem.

**Comment by atom992** [comment]
My app is based on react-native too. @npomfret  which version of react-native you used? 
I am using react-native 0.17.0

**Comment by hideki** [comment]
Hi @atom992 and @npomfret,

Currently we don't know what causes this problem. It seems `SyncGatewaySession` value is valid which was verified by the curl command. The request from CBL does not work, but the request from curl command works. 

I wonder if SG might not receive `SyncGatewaySession` value from CBL client when authentication fails. Could you please enable `Bucket`logging at Sync Gateway side? If SG receives  `SyncGatewaySession` value, SG prints like: `Bucket: Get(_sync:session:27ffa471901409fc85c64ea541af7acdc071ee86) [21ms]`. If SG does not receive `SyncGatewaySession` value, SG does not print anything.  

I hope this give us more hints.

Thanks!

**Comment by atom992** [comment]
ok.

**Comment by npomfret** [comment]
I'm using 0.23.1, and believe the problem happens on both IOS and Android.  Retrying the request works around the issue for the moment.  To be clear, I experience the 401 problem using the REST api in CBL.   I do not have problems talking to the sync gateway.

**Comment by jamiltz** [comment]
I'm using 0.23.1, and believe the problem happens on both IOS and Android.
Interesting! Could it be an issue with RN?

I recommend using Charles for network debugging the CBL REST API on a device or emulator. Start Charles then from the Android WiFi page, select the current network and add a manual proxy in advanced settings (set the hostname to the IP of your computer and port to 8888 - the default reverse proxy port on Charles). Return to the app and Charles should pick up the traffic going to the CBL Listener. See if the 401 occurs because the auth credentials are missing in the request.

Retrying the request works around the issue for the moment. To be clear, I experience the 401 problem using the REST api in CBL. I do not have problems talking to the sync gateway.
Looks like this issue was opened for sync gateway sessions which work differently to authentication on the CBL REST API. Feel free to open a new issue for this scenario if you manage to pin down whether it comes from CBL or RN.

**Comment by jamiltz** [comment]
> I'm using 0.23.1, and believe the problem happens on both IOS and Android.

Could it be an issue with RN?

I recommend using Charles for network debugging the CBL REST API on a device or emulator. Start Charles then from the Android WiFi page, select the current network and add a manual proxy in advanced settings (set the hostname to the IP of your computer and port to 8888 - the default reverse proxy port on Charles). Return to the app and Charles should pick up the traffic going to the CBL Listener. See if the 401 occurs because the auth credentials are missing in the request.

> Retrying the request works around the issue for the moment. To be clear, I experience the 401 problem using the REST api in CBL. I do not have problems talking to the sync gateway.

Looks like this issue was opened for sync gateway sessions which work differently to authentication on the CBL REST API. Feel free to open a new issue for this scenario if you manage to pin down whether it comes from CBL or RN.

**Comment by hideki** [comment]
Hi all,
Do you have any additional information about this ticket? If not, I'd like to close this ticket.
Thanks!

**Comment by hideki** [comment]
As we don't have additional information to investigate this issue further, so I will close this ticket. Please reopen the ticket if you still encounter this issue.

**Comment by npomfret** [comment]
I think I'm able to reproduce this now.  First I get my app running in which it has 2 replication tasks.  I then delete all sessions in the system using the admin rest api, and then restart my android app.  The app then creates a new sessions and then attempts to create 2 replication tasks...

The new session that got created looks like this:

`SyncGatewaySession=d92f44b908b5c9dd20b351fb4b8a6b273433387b; Path=/sync_gateway; Expires=Thu, 11 Aug 2016 17:53:49 GMT
`

I've tested it with the SG admin api (`/sync_gateway/_session/d92f44...`) and it returns with `ok: true`.

But shortly after my android app gets a 401.  Here's the failing request in the debugger:

![screen shot 2016-08-10 at 19 00 05](https://cloud.githubusercontent.com/assets/5489469/17565129/fcaa2fb4-5f2d-11e6-89ef-b1991e30a5f9.png)

I don't recognise the url (`.../sync_gateway/_local/496b46cec1f038801151ff4bf13edf57c8abed82`), here, but I think my app was creating a replication session when this happened.

And in the SG logs I see a corresponding:

```
18:55:01.822454 2016-08-10T18:55:01.822+01:00 HTTP:  #018: GET /sync_gateway/_session/d92f44b908b5c9dd20b351fb4b8a6b273433387b  (ADMIN)
18:55:01.823955 2016-08-10T18:55:01.823+01:00 HTTP+: #018:     --> 200   (1.5 ms)
18:55:12.706641 2016-08-10T18:55:12.706+01:00 HTTP:  #019: GET /sync_gateway/_local/496b46cec1f038801151ff4bf13edf57c8abed82
18:55:12.706705 2016-08-10T18:55:12.706+01:00 HTTP: #019:     --> 401 Login required  (0.5 ms)
```

So (I hope) you can see from the debug information that the correct session id (`d92f44...`) is being used in the request, but the SG is still returning a 401.

If I delete my entire Android database I can start replication tasks again.  It's as if theres something in the Android db that's somehow linked to the dead session.

Using 1.3.0-45. Don't believe it happens in the latest iOS build.

**Comment by pasin** [comment]
I noticed from your screenshot that the Cookie value`SyncGatewaySession=d9....7b;Path=`) in the request header includes path and expiration date information. This looks like the whole cookie string, not just the key=value. I don't know if this is correct and causes the issue or not.

**Comment by npomfret** [comment]
Disappointed to see 1.3 being released without getting this resolved.  What happens when a sessions expires?  Is it possible to get replication working again without deleting the entire database?

**Comment by pasin** [comment]
@hideki I think this issue should be carries over to 1.3.1. I will get this to review to include in 1.3.1.

**Comment by npomfret** [comment]
So am I right in thinking that any app written with v1.3.0 that uses replication will stop working after 24 hours because the session will expire and there's no way to create a 2nd replication task with a different session?

**Comment by pasin** [comment]
@npomfret I couldn't reproduce the issue with couchbase-lite-android-lite-serv 1.3.0 with the steps you described. I tested both push/pull and single-shot/continuous. A new re-created session cookie set to the POST /_replicate header works correctly.

> If I delete my entire Android database I can start replication tasks again. It's as if theres something in the Android db that's somehow linked to the dead session.
 We do have a cookie storage which persists cookies in the database internally. However there shouldn't be any cookies set into that cookie storage as the replicators started by POST /_replicate not have any authenticator set so the login logic that will create a new SGW session will be bypassed.
 
@npomfret how do you detect session expiration and restart the SGW? In 1.3.1, we fixed a couple of issue around POST /_replicate, I don't know if that has any contribution to the issue that you are encountering or not. If I give you a 1.3.1 candidate binary, are you able to try it out?

**Comment by pasin** [comment]
Another thing that you could try is removing this portion `Path=/sync_gateway; Expires=Thu, 11 Aug 2016 17:53:49 GMT` when you set the cookie in the header and see if that helps or not.

**Comment by pasin** [comment]
Also like @jamiltz's recommendation, it might be useful to see the raw request to SGW when the issue happens. We might spot something from that. Normally I use httpscoop to capture the requests, but Charles should work as well.

**Comment by pasin** [comment]
A couple of things to try:
1. Remove Path and Expires info from the cookie header and see if it helps or not.
2. Use Charles or HTTPSCoop to capture the RAW request that has 401 error. This is the best way to see if the request actually has the cookie set or not. If the request is correct, we may need to look the issue on the SGW side.

Please let me know. Note that I edited and deleted some of my previous comments above as I think these two items are best to help investigating the issue.

**Comment by npomfret** [comment]
What steps did you take to reproduce?  You need to:

1) Create a session on a client
2) Set up bi-directional replication on the client
3) Delete the session on the server using the REST API while the client is running (to simulate it expiring)
4) Attempt to create another session on the client after it detects that its current session isn't valid

**Comment by pasin** [comment]
@npomfret, Those are the steps I tried to reproduce **except** the step 4 that I didn't implement any logic to detect the invalid session. I just cancelled both continuous push and pull replicator after step 3 and restarted with a new session cookies.

Could you please capture the RAW request of the request that has 401 error? This will help us know for sure if the request is sent correctly with the right cookie in the request header.

**Comment by pasin** [comment]
@npomfret do you have any updates? We are closing down the 1.3.1 release. It will be great if you could provide more information as requested above asap. Please let me know.

**Comment by npomfret** [comment]
I'll get tested 1.3.1 today

**Comment by npomfret** [comment]
Hi.  I re-tested with 1.3.1-16 and some good news, which is that the problem as I described is indeed not reproducible any more. However, I did encounter another change in behaviour that I wasn't expecting.

The log below is from my app.  It was running at the moment I deleted all user sessions from the server using the admin rest api.  I typed into the log 'done' at the point they were deleted (around 08-23 11:43:47.727) so you could see when things were happening.

My app checks the replication tasks every 30 seconds to see if there are any problems.  I expected them to be in a state of 'STOPPED' because the session they are using is not longer valid.  But nothing changed.  You can see from my apps logging that they remain in a state of idle.

To show that the session is invalid my app then tries to use it to write directly to the sync gateway via the REST API, but you can see it gets back a `{ error: 'Unauthorized', reason: 'Login required' }`. 

Here's the log:
https://gist.github.com/npomfret/ae5978abdf321e6e9f70421f8efa7699

I guess because of this change my app is no longer able to detect dead sessions, so the test i wanted to run wasn't possible.  However, after restarting it and forcing it to create a new session it was fine - which certainly was not the case before. 

I'm going to have to modify the app to test the session against the sync-gateway, and if the test fails I'll force it to re-create the replication tasks.  I will let you know the results later.

What is the best URL to hit (with lowest overhead) to test if a session is working?  Can I just hit the root url?

**Comment by npomfret** [comment]
Ok, so i changed my app to **not** use the state of the replication tasks to detect if they were working, because as stated above, they don't seem to update to 'stopped' when the user session gets deleted.  

Instead, my app now uses the sync-gateway REST API to explicitly check the session periodically, and if its invalid, then create a new session and restart the replication tasks using the new session.  And it appears to work perfectly.  Here's the log:

https://gist.github.com/npomfret/dd56c43accb2f4e6ef50759d204b8363

Thanks.

**Comment by pasin** [comment]
@npomfret Thanks for testing and getting back to us so quickly.

> My app checks the replication tasks every 30 seconds to see if there are any problems. I expected them to be in a state of 'STOPPED' because the session they are using is not longer valid. But nothing changed. You can see from my apps logging that they remain in a state of idle.

The replicator cannot automatically recognized the session expiration until the replicator tries to communicate with SGW with the old session cookies. The continuous pull replicator will even need more time as it cannot recognize until the _change feeds is timeout (See discussion in couchbase/couchbase-lite-java-core#1397). I guessed if you updated any documents on the client side, the push replicator should report the 401 error and the _active_tasks result should reflect that. I will update https://github.com/couchbaselabs/mobile-testkit/issues/641 to test this case.

I think as of the current API and implementation, it's simple and effective enough to check the invalid session by sending a sort of ping request SGW directly. Sending to the root URL should has the lowest overhead.

I'm closing the issue now. Thanks!

**Comment by hideki** [comment]
I believe this issue is same problem with https://github.com/couchbase/couchbase-lite-java-core/issues/1617
